### PR TITLE
Add logging support

### PR DESCRIPTION
**Pull request by alexdewar**
I finally decided to add some logging support to BoB robotics (issue #10 ) mostly because I was getting fed up of writing ad hoc code to print various bits of information when I was debugging.

I've gone with plog, because it seems a sensible choice (and I think you're also using it for GeNN?).

### CODE REVIEW

**Comment by alexdewar** [comment]
Hey @neworderofjamie , would you mind taking a quick look at this? It shouldn't break anything and would be nice to merge (once CI is done running)...

**Comment by neworderofjamie** [comment]
before I forget and on the subject of CI, thanks to Stefan, we now have a windows node. Once again, thanks to GeNN, I've figured out the unpleasant Jenkins hoops to jump through to use it so would you like me to add Windows building to the BoB robotics Jenkinsfile?

**Comment by alexdewar** [comment]
Yes please! That would be super handy.

The only thing is that I guess you're building using msbuild and VS projects (?) and our VS solution for the examples is currently a bit sparse. It's worth doing in any case, but I suppose it would also be good to add the other examples to the solution at some future date so we have better coverage.

⁣Sent from BlueMail ​

**Comment by neworderofjamie** [comment]
Yeah - you can just use msbuild from the command line to be build all
targets in a solution

**Comment by alexdewar** [comment]
Cool. Something else that's been on my to-do list for CI is to automatically build-test the tools and projects too. There are build scripts in the folders... Would you mind taking a look at that if it's not too much effort? I had a go but failed miserably with Scala. Obvs not a priority though!

⁣Sent from BlueMail ​

**Comment by alexdewar** [comment]
I am also wondering if we've reached the stage where cmake might be a better option for us cf. recursive makefiles etc. It is good for building subprojects and would allow us to target Windows and Linux easily....

⁣Sent from BlueMail ​

**Comment by neworderofjamie** [comment]
No worries - as my one line changes and 4-letter commit messages attest -
it's a massive pain :)

File: common/logging.h [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        /*
+         * Also print log messages to console in colour.
+         *
+         * The plog documentation reports that this should have static scope.
+         */
+        static plog::ColorConsoleAppender<Formatter> appender;
+        plog::get()->addAppender(&appender);
+    }
+
+    static constexpr auto logLevelToString(plog::Severity severity)
```
Comment (neworderofjamie) [inline common/logging.h]: Can you reuse the versions here https://github.com/SergiusTheBest/plog/blob/365646171e192d60fa1c2509d6a20adb5893e851/include/plog/Severity.h#L16-L49?

File: common/logging.h [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+        throw std::runtime_error(std::string(severityEnvVar) + " is not a valid log level");
+    }
+}; // Logger
+
+/*
+ * We use this dummy object with global scope so that logging is initialised
+ * before main() is called.
+ */
+#ifndef NO_HEADER_DEFINITIONS
```
[the remark this answers is not in the document]
Comment (alexdewar) [reply common/logging.h]: Good idea. C++17 seems to have a few extra things that are quite nice and it looks like it should work on the gcc shipped with Ubuntu 16.04 (although annoyingly it looks like ``std::filesystem`` wasn't added until gcc v8)...

File: common/logging.h [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+{
+public:
+    static Logger &getInstance()
+    {
+        static Logger logger;
+        return logger;
+    }
+
+private:
+    //! Plog's default TxtFormatter is a bit verbose, so let's implement our own
```
[the remark this answers is not in the document]
Comment (alexdewar) [reply common/logging.h]: Cool. Maybe if we use (roughly) the same header in GeNN we should add another header guard to stop the plog headers being included twice? (i.e. ``#define GENN_LOGGING`` or something)

File: navigation/infomax.h [hunk 4]
```diff
@@ -132,32 +131,26 @@ class InfoMax : public VisualNavigationBase
             }
         }
 
-        // std::cout << "Initial weights" << std::endl
-        //           << weights << std::endl;
+        // LOG_INFO << "Initial weights" << weights;
```
[the remark this answers is not in the document]
Comment (alexdewar) [reply navigation/infomax.h]: Good idea

File: robots/tank_pid.h [hunk 5]
```diff
@@ -91,7 +92,7 @@ class TankPID
         switch (state) {
         case TankPIDState::OrientingToGoal:
             if (event == Event::Enter) {
-                std::cout << "Starting turning" << std::endl;
+                LOG_INFO << "Starting turning";
```
[the remark this answers is not in the document]
Comment (alexdewar) [reply robots/tank_pid.h]: :+1:

File: common/logging.h [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+    static plog::Severity getLogLevel()
+    {
+        // Try to get from environment variable
+        char *severityEnvVar = std::getenv("LOG_LEVEL");
+        if (!severityEnvVar) {
+            return DefaultLogLevel;
+        }
+
+        // Convert to uppsercase
+        for (char *c = severityEnvVar; *c; c++) {
```
[the remark this answers is not in the document]
Comment (alexdewar) [reply common/logging.h]: The only thing is that it's a C-string so to use ``std::transform`` we'd first have to work out the length.

File: common/logging.h (hunk 1 above)
Comment (alexdewar) [reply common/logging.h]: I did that initially but I had a couple of issues with those versions:
1. The string names didn't seem too obvious to me (e.g. ``VERB`` for verbose)
2. If you pass in some random string it'll return a log level of none, whereas I thought it'd be better to give an error (for e.g. if you make a typo)

**Comment by alexdewar** [comment]
Ta for looking at this! All helpful suggestions.

File: common/logging.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (alexdewar) [reply common/logging.h]: It also looks like Ubuntu 18.04 uses gcc v7 so I guess that means no ``std::filesystem`` until 20.04!

**Pull request merged**